compiler/src/dotty/tools/dotc/core/SymbolLoaders.scala

```diff
@@ -333,7 +333,15 @@ abstract class SymbolLoader extends LazyType { self =>
     def description(using Context): String = s"proxy to ${self.description}"
   }
 
-  override def complete(root: SymDenotation)(using Context): Unit = {
+  private inline def profileCompletion[T](root: SymDenotation)(inline body: T)(using Context): T = {
+    val sym = root.symbol
+    def associatedFile = root.symbol.associatedFile match
+      case file: AbstractFile => file
+      case _ => NoAbstractFile
+    ctx.profiler.onCompletion(sym, associatedFile)(body)
+  }
+
+  override def complete(root: SymDenotation)(using Context): Unit = profileCompletion(root) {
```

Comment:
I checked that the `associatedFile` callback above doesn't run when profiling is disabled ✅ 

But I am still not 100% convinced that is zero cost—it seems that calls to default implementations of `beforeCompletion` and `afterCompilation` will still exist after inlining? This would just be 2 method calls and a pair allocation, so maybe harmless in terms for performance?

Unfortunately, the benchmarks bot is currently down, so we can't run the benchmarks now.

Reply:
The allocations have been reduced by preallocating empty (default) outputs. This should reduce any overhead when -Yprofile-trace is disabled. The 2 additional calls to before/after completion might still be present, but maybe JVM can optimize these? 

Do you maybe know if the benchmarks bot is working now? We could run it now, after the rebase, to ensure the overhead is not significant when profile-tracing is disabled.